lib/mongo/auth/scram256/conversation.rb

```diff
@@ -0,0 +1,509 @@
+        #
+        # @since 2.6.0
+        def client_first_message
+          BSON::Binary.new("n,,#{first_bare}")
+        end
+
+        # Client final implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-7
+        #
+        # @since 2.6.0
+        def client_final
+          @client_final ||= client_proof(client_key, client_signature(stored_key(client_key), auth_message))
+        end
+
+        # Client key algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-3
+        #
+        # @since 2.6.0
+        def client_key
+          @client_key ||= hmac(salted_password, CLIENT_KEY)
+          user.instance_variable_set(:@client_key, @client_key) unless user.send(:client_key)
+          @client_key
+        end
+
+        # Client proof algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-3
+        #
+        # @since 2.6.0
+        def client_proof(key, signature)
+          @client_proof ||= Base64.strict_encode64(xor(key, signature))
+        end
+
+        # Client signature algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-3
+        #
+        # @since 2.6.0
+        def client_signature(key, message)
+          @client_signature ||= hmac(key, message)
+        end
+
+        # First bare implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-7
+        #
+        # @since 2.6.0
+        def first_bare
+          @first_bare ||= "n=#{user.encoded_name},r=#{nonce}"
+        end
+
+        # H algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-2.2
+        #
+        # @since 2.6.0
+        def h(string)
+          digest.digest(string)
+        end
+
+        # HI algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-2.2
+        #
+        # @since 2.6.0
+        def hi(data)
+          OpenSSL::PKCS5.pbkdf2_hmac(
+            data,
+            Base64.strict_decode64(salt),
+            iterations,
+            digest.size,
+            DIGEST
+          )
+        end
+
+        # HMAC algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-2.2
+        #
+        # @since 2.6.0
+        def hmac(data, key)
+          OpenSSL::HMAC.digest(digest, data, key)
+        end
+
+        # Get the iterations from the server response.
+        #
+        # @api private
+        #
+        # @since 2.6.0
+        def iterations
+          @iterations ||= payload_data.match(ITERATIONS)[1].to_i.tap do |i|
+            next unless i < MIN_ITER_COUNT
```

Comment:
This is a very strange construct, can the block body be rephrased to `if i < MIN_ITER_COUNT then raise ...` ?

Reply:
I think we should hold off on refactoring the conversation algorithm for another time (and more broadly, not bring extra scope by reviewing existing code that works fine when doing a review). I understand that being newly introduced to the codebase, you're going to notice a lot of small things that look like they can be improved, and in a lot of places, that's probably correct. However, most of these things can a) be addressed in a separate ticket/commit and b) are not a huge priority while we're driving towards getting a release in time to support 4.0. There's certainly a lot of work to be done in cleaning up the code in a lot of places, but it's important that we don't get sidetracked by issues in existing code that is working properly. In the future, I'll try to do a better job of marking which code in a pull request is not new so that we can focus on the new stuff being added.

Reply:
Oops, I completely misread this; this is clearly not existing code, so you can ignore what I said. Sorry about that!